### PR TITLE
Add support for PyPy

### PR DESCRIPTION
**Pull request by brouberol**
Cliquet now supports PyPy, by:

* using ``json`` instead of ``ujson`` (with PyPy only), because ``usjon`` is written in C and uses the CPython C API
* using ``psycopg2cffi`` instead of ``psycopg2`` (with PyPy only)

Warning: this is still a work in progress.
- [x] needs JSONB support in ``psycopg2cffi`` (see https://github.com/chtd/psycopg2cffi/issues/36), or a way to handle deserializing JSONB fields
- [x] a mention of pypy support in the documenation

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,9 +26,12 @@
     'requests',
     'six',
     'structlog',
-    'ujson',
 ]
 
+# ujson is not pypy compliant, as it uses the CPython C API
+if not installed_with_pypy:
```
Comment (Natim) [inline setup.py]: If we do that for ``ujson``, we can probably do it as well for ``psycopgcffi`` in the ``cliquet[postgresql]`` rule

**Comment by brouberol** [comment]
I have been thinking about the suggested implementation, and there's something that bugs me: the interpreter has an effect on the dependencies. If that was only about ``ujson`` not being installable through PyPy, that'd probably be ok. For me, the real issue is the ``psycopg2/psycopg2cffi`` dichotomy. 

We've seen that ``psycopg2cffi`` does not have the same development rate that ``psycopg2``, as some features introduced in PostgreSQL 9.2 (and ``psycopg2`` ~ a year ago -- see https://github.com/psycopg/psycopg2/issues/226) were not yet reflected in ``psycopg2cffi``. My main fear is that this would introduce some maintainability issues between CPython and PyPy.

So I guess what I'm really asking is, why do we want to support PyPy, what is the real gain, and is it worth the risk of introducing two different Postgres python drivers with different development rates?   

My 0.02$.

**Comment by Natim** [comment]
This is interesting. Also we have tests that make sure that the backend does what we want it to do.
So if test are passing with PyPy and psycopg2cffi we can be confident that our storage works with PyPy and psycopg2cffi

**Comment by brouberol** [comment]
So, now that JSONB support landed in psycopg2cffi, all tests are green for the PyPy tox target. I'll wait for a PyPI release and include a ``psycopg2cffi<=VERSION`` clause in the setup.py.

I'll also add some documentation about PyPy support, and the fact that it uses a different python driver for Postgres. 

Also, happy Canada 

Canada!

**Comment by brouberol** [comment]
Side note: we need to wait for the PyPI release of the new psycopg2cffi (with JSONB support) before merging this one.

File: setup.py [hunk 2]
```diff
@@ -24,14 +26,22 @@
...
 
-DEPENDENCY_LINKS = [
-]
+if not installed_with_pypy:
+    # ujson is not pypy compliant, as it uses the CPython C API
+    REQUIREMENTS.append('ujson')
+    # We install psycopg2cffi instead of psycopg2 when dealing with pypy
+    # Note: JSONB support landed after psycopg2cffi 2.7.0
+    POSTGRESQL_REQUIRES = [
+        'psycopg2cffi>2.7.0',
```
Comment (leplatrem) [inline setup.py]: Isn't it the opposite?

File: cliquet/utils.py [hunk 3]
```diff
@@ -7,7 +7,21 @@
...
 
-import ujson as json  # NOQA
+# ujson is not installable with pypy
+try:  # pragma: no cover
+    import ujson as json  # NOQA
+except ImportError:  # pragma: no cover
+    import json  # NOQA
+
+# psycopg2cffi is installed under pypy, instead of psycopg2
+# pragma: no cover
```
Comment (leplatrem) [inline cliquet/utils.py]: Some pragmas can be removed IMO

Comment (brouberol) [inline cliquet/utils.py]: You mean removing the "pragma: no cover" in the "except" clause?

**Comment by leplatrem** [comment]
Very good job! Thanks!

I'm sorry that I had put it as easy-pick... Was not so trivial to tackle :/

File: cliquet/utils.py (hunk 3 above)
Comment (leplatrem) [inline cliquet/utils.py]: Yes, the ones in front of try (always executed unlike except)

**Comment by brouberol** [comment]
Heh, that was a fun one anyway.

**Comment by Natim** [comment]
Good Job @brouberol and thanks you for the fish :)

**Pull request merged**